### PR TITLE
ICR

### PR DESCRIPTION
**Pull request by as2388**
Review #3 first

### CODE REVIEW

File: project_redss/auto_code_show_messages.py [hunk 1]
```diff
@@ -1,65 +1,86 @@
...
+            output_path = path.join(coda_output_dir, "{}.json".format(plan.coda_filename))
+            with open(output_path, "w") as f:
+                TracedDataCoda2IO.export_traced_data_iterable_to_coda_2(
+                    not_noise, plan.raw_field, cls.SENT_ON_KEY, plan.id_field, {}, f
+                )
+
+        # Output messages for ICR
+        IOUtils.ensure_dirs_exist(icr_output_dir)
+        for plan in DatasetSpecification.RQA_CODING_PLANS:
+            possible_messages = [msg for msg in not_noise if "ControlCode" not in msg.get(plan.coded_field, [{}])[0]]
```
Comment (as2388) [inline project_redss/auto_code_show_messages.py]: will rewrite this line

File: project_redss/auto_code_show_messages.py [hunk 2]
```diff
@@ -72,16 +71,22 @@ def auto_code_show_messages(cls, user, data, icr_output_dir, coda_output_dir):
...
-        # IOUtils.ensure_dirs_exist_for_file(icr_output_path)
-        with open(icr_output_dir, "w") as f:
-            f.write("")
-            # TracedDataCSVIO.export_traced_data_iterable_to_csv(icr_messages, f, headers=[run_id_key, raw_text_key])
+        # Output messages for ICR
+        IOUtils.ensure_dirs_exist(icr_output_dir)
+        for plan in DatasetSpecification.RQA_CODING_PLANS:
+            rqa_messages = []
+            for td in not_noise:
+                # This test works because the only codes which have been applied at this point are TRUE_MISSINGs.
```
Comment (lukechurch) [inline project_redss/auto_code_show_messages.py]: Please assert this

File: project_redss/auto_code_show_messages.py [hunk 3]
```diff
@@ -72,16 +71,22 @@ def auto_code_show_messages(cls, user, data, icr_output_dir, coda_output_dir):
...
+                # If any other coding is done above, this test will need to change.
+                if plan.coded_field not in td:
+                    rqa_messages.append(td)
+
+            icr_messages = ICRTools.generate_sample_for_icr(rqa_messages, cls.ICR_MESSAGES_COUNT, random.Random(0))
```
Comment (lukechurch) [inline project_redss/auto_code_show_messages.py]: Promote the random seed to a defined constant in the file?

File: project_redss/auto_code_show_messages.py [hunk 4]
```diff
@@ -72,16 +71,22 @@ def auto_code_show_messages(cls, user, data, icr_output_dir, coda_output_dir):
...
+
+            icr_output_path = path.join(icr_output_dir, f"{plan.coda_filename}.csv")
```
Comment (lukechurch) [inline project_redss/auto_code_show_messages.py]: {coda_filename}.csv?

Is this a legacy from the days when coda took csv files?

**Review by lukechurch — APPROVED** [review]
With clarification on the variable names, I guesssss I'm ok with this. Let's file a bug against the pipeline so that when we copy it forwards to the next project this is one of the things we fix.

**Pull request merged**